docs/document.py

```diff
@@ -0,0 +1,290 @@
+import glob
+import json
+import os
+import pathlib
+import re
+from typing import Union
+
+import pandas as pd
+
+
+def _format_cell_by_type(x, header):
+    if "enum" in header and len(x) > 1:
+        return "Allowed Values: `" + ",".join(map(str, x)) + "`"
+    if header == "foreign_key" and len(x) > 1:
+        table, var = x.split(".")
+        if table == "":
+            return "Variable: `{}`".format(var)
+        return "Table: `{}`, Variable: `{}`".format(table, var)
+    else:
+        return str(x)
+
+
+def _recursive_items(d: dict, key_prefix: str = ""):
+    """
+    Recursively flattens a nested dictionary and returns a dictionary with key_prefixe
+    to represent original nesting structures.
+    Modified from https://stackoverflow.com/questions/39233973/get-all-keys-of-a-nested-dictionary
+    """
+    for k, v in d.items():
+        if type(v) is dict:
+            yield from _recursive_items(v, key_prefix=k + " ")
+        elif k == "required":
+            yield (k, v)
+        else:
+            yield (key_prefix + k, v)
+
+
+def _fill_template(
+    outfile_path: Union[str, pathlib.Path],
+    content_dict: dict,
+    template_path: Union[str, pathlib.Path] = None,
+) -> None:
+    """Replace contents of a text file with template variables filled by a dictionary.
+
+    Args:
+        outfile_path (Union[str,pathlib.Path]): File to write output to.
+        content_dict (dict[): Dictionary of values to fill template with.
+        template_path (Union[str,pathlib.Path], optional): _description_. If not specified, will assume
+            it is the outfile with a .template suffix (i.e. index.template.md)
+
+    Raises:
+        ValueError: _description_
+    """
+
+    outfile_path = pathlib.Path(outfile_path)
+    if template_path is None:
+        template_path = outfile_path.parent / pathlib.Path(
+            outfile_path.stem + ".template" + outfile_path.suffix
+        )
+
+    with open(template_path) as _template_file:
+        _txt_contents = _template_file.read().format(**content_dict)
+        with open(os.path.join(outfile_path), "w") as _outfile:
+            _outfile.write(_txt_contents)
+
+
+def _list_to_md_table(list_of_dicts: list) -> str:
+    """
+    Reads a list of dictionaries defining a schema and creates a markdown table.
+    args:
+        list_of_dicts: a list of dictionaries containing definition of fields.
+    returns: A markdown string representing the table.
+    """
+    # flatten dictionary
+    flat_list = []
+    for i in list_of_dicts:
+        flat_list.append({k: v for k, v in _recursive_items(i)})
+
+    # get all items listed
+    standard_header_items = [
+        "name",
+        "required",
+        "type",
+        "foreign_key",
+        "description",
+    ]
```

Comment:
The tables produced are not great!

- foreign_key field is always empty because it's a table property, not a field in the 'fields' object/dict.
- would be nice to have one constraints field that listed all constraints, and converted the required constraint to "Required" when true, or "Optional" if false. For example:
   ```
    | field | constraints |
    |---|---|
    | device_id | Required, Unique |
    ```
- line breaks `\n` from the spec within table cells break table formatting, need to be converted to `<br />`.

Reply:
> The tables produced are not great!

I agree!  Happy to iterate.  This was a first try to get *something* up. Happy to have others help/suggest fixes. I'll try your suggestions above for now.

Reply:
This looks much better! I'm happy with this as a starting place.